votor/src/consensus_pool/vote_pool.rs

```diff
@@ -1,129 +1,304 @@
+//! Container to store received votes and associated stakes.
+//!
+//! Also implements various checks for slashable behavior and duplicate votes.
+
 use {
-    crate::common::Stake,
-    agave_votor_messages::consensus_message::VoteMessage,
+    crate::common::{Stake, VoteType},
+    agave_votor_messages::{consensus_message::VoteMessage, vote::Vote},
+    solana_clock::Slot,
     solana_hash::Hash,
     solana_pubkey::Pubkey,
-    std::collections::{BTreeMap, BTreeSet},
+    std::collections::{btree_map::Entry, BTreeMap},
+    thiserror::Error,
 };
 
-/// There are two types of vote pools:
-/// - SimpleVotePool: Tracks all votes of a specfic vote type made by validators for some slot N, but only one vote per block.
-/// - DuplicateBlockVotePool: Tracks all votes of a specfic vote type made by validators for some slot N,
-///   but allows votes for different blocks by the same validator. Only relevant for VotePool's that are of type
-///   Notarization or NotarizationFallback
-pub(super) enum VotePool {
-    SimpleVotePool(SimpleVotePool),
-    DuplicateBlockVotePool(DuplicateBlockVotePool),
+#[derive(Debug, PartialEq, Eq, Error)]
+pub(crate) enum AddVoteError {
+    #[error("duplicate vote")]
+    Duplicate,
+    #[error("slashable behavior")]
+    Slash,
 }
 
-#[derive(Default)]
-pub(super) struct SimpleVotePool {
-    votes: Vec<VoteMessage>,
-    total_stake: Stake,
-    prev_voted_validators: BTreeSet<Pubkey>,
+/// Container to store per slot votes.
+struct Votes {
+    /// The slot this instance of Votes is responsible for.
+    slot: Slot,
+    /// Skip votes are stored in map indexed by validator.
+    skip: BTreeMap<Pubkey, VoteMessage>,
+    /// Skip fallback votes are stored in map indexed by validator.
+    skip_fallback: BTreeMap<Pubkey, VoteMessage>,
+    /// Finalize votes are stored in map indexed by validator.
+    finalize: BTreeMap<Pubkey, VoteMessage>,
+    /// Notar votes are stored in map indexed by validator.
+    notar: BTreeMap<Pubkey, VoteMessage>,
+    /// A validator can vote notar fallback on upto 3 blocks.
+    ///
+    /// Per validator, we store a map of which block ids the validator has voted notar fallback on.
+    notar_fallback: BTreeMap<Pubkey, BTreeMap<Hash, VoteMessage>>,
```

Comment:
I think the general recommendation is for <=3 entries, Vec is actually more efficient than BTreeMap because of the overhead of BTreeMap.

Reply:
A vec does not provide us with a guarantee that we do not have duplicates.  And for small enough sets, the performance of both containers is comparable.  And I believe that we are using BTree* in other places instead of Vec for small sets as well.